fpm/src/fpm_compiler.f90

```diff
@@ -0,0 +1,224 @@
+       & -fPIC&
+       & -fmax-errors=1&
+       & -ffast-math&
+       & -funroll-loops&
+       & -fcoarray=single&
+       &'
+    case('debug_gfortran')
+       module_path_switch='-J '
+       fflags = '&
+       & -Wall&
+       & -Wextra&
+       & -Wimplicit-interface&
+       & -fPIC -fmax-errors=1&
+       & -g&
+       & -fbounds-check&
+       & -fcheck-array-temporaries&
+       & -fbacktrace&
+       & -fcoarray=single&
+       &'
+
+    case('release_f95')
+       module_path_switch='-J '
+       fflags='&
+       & -O3&
+       & -Wimplicit-interface&
+       & -fPIC&
+       & -fmax-errors=1&
+       & -ffast-math&
+       & -funroll-loops&
+       &'
+    case('debug_f95')
+       module_path_switch='-J '
+       fflags = '&
+       & -Wall&
+       & -Wextra&
+       & -Wimplicit-interface&
+       & -fPIC -fmax-errors=1&
+       & -g&
+       & -fbounds-check&
+       & -fcheck-array-temporaries&
+       & -Wno-maybe-uninitialized -Wno-uninitialized&
+       & -fbacktrace&
+       &'
+
+    case('release_nvfortran')
+       module_path_switch='-module '
+       fflags = '&
+       & -Mbackslash&
+       &'
+    case('debug_nvfortran')
+       module_path_switch='-module '
+       fflags = '&
+       & -Minform=inform&
+       & -Mbackslash&
+       & -g&
+       & -Mbounds&
+       & -Mchkptr&
+       & -Mchkstk&
+       & -traceback&
+       &'
+
+    case('release_ifort')
+       module_path_switch='-module '
+       fflags = '&
+       & -fp-model precise&
+       & -pc 64&
+       & -align all&
+       & -coarray&
+       & -error-limit 1&
+       & -reentrancy threaded&
+       & -nogen-interfaces&
+       & -assume byterecl&
+       & -assume nounderscore&
+       &'
+    case('debug_ifort')
+       module_path_switch='-module '
+       fflags = '&
+       & -warn all&
+       & -check:all:noarg_temp_created&
+       & -coarray&
+       & -error-limit 1&
+       & -O0&
+       & -g&
+       & -assume byterecl&
+       & -traceback&
+       &'
+    case('release_ifx')
+       module_path_switch='-module '
+       fflags = ' '
+    case('debug_ifx')
+       module_path_switch='-module '
+       fflags = ' '
+
+    case('release_pgfortran','release_pgf90','release_pgf95')  ! Portland Group F90/F95 compilers
+       module_path_switch='-module '
+       fflags = ' '
+    case('debug_pgfortran','debug_pgf90','debug_pgf95')  ! Portland Group F90/F95 compilers
+       module_path_switch='-module '
+       fflags = ' '
+
+    case('release_flang')
+       module_path_switch='-module '
+       fflags = ' '
+    case('debug_flang')
+       module_path_switch='-module '
+       fflags = ' '
+
+    case('release_lfc')
+       module_path_switch='-M '
+       fflags = ' '
+    case('debug_lfc')
+       module_path_switch='-M '
+       fflags = ' '
+
+    case('release_nagfor')
+       module_path_switch='-mdir '
+       fflags = ' '
+    case('debug_nagfor')
+       module_path_switch='-mdir '
+       fflags = ' '
```

Comment:
I was looking at adding nag support to the bootstrap version, and was going to use `-O4` for release, and `-g -Call` for debug. On a related note, we've got `-fPIC` for gfortran, should that be included for all compilers that support it?

Reply:
Looking for any suggestions for what the flags should be and still want to reimplement the loader and --flags options in some form.  Was waiting for this PR as do not want to built up too many diffs as the --install switch changes some of the same files.  That seems like a good change. Many of these are still blank. The -fPIC was in the original but  I personally think that is a good suggestion to add it where available.  I will add a matching change for `nagfor`.  The functionality is there in this PR (I have been using it with three compilers).  I would like to add short names like -c||-C or even -e (for encoder)  or -F for Fortran for --compiler and -V for verbose but this PR is already a little top heavy. Sounds good. Was hoping for a lively discussion on what options to add for some of the compilers I do not have access to, but have not seen much. Hopefully you will spur a discussion on -fPIC and maybe some other options. Right now without fpm.toml supporting option selection and more arguably not having support for a custom build script and CLI options for compiling and loading getting good defaults is especially important. Some if personal taste (I dislike the default allowing C-like escape characters, for example).

Reply:
My guess is the following, but I have no access to test it.

```fortran   case('release_nagfor')
       module_path_switch='-mdir '
       fflags = ' &
       & -O4&
       & -coarray=single&
       & -PIC&
       '
    case('debug_nagfor')
       module_path_switch='-mdir '
       fflags = '&
       & -g&
       & -C=all&
       & -O0&
       & -gline&
       & -coarray=single&
       & -PIC&
       '
```

Reply:
NAG will only output modules in the selected directory, but not read from there, I had to apply a quick hack to do a test with `nagfor`:

```diff
diff --git a/fpm/src/fpm_compiler.f90 b/fpm/src/fpm_compiler.f90
index 540b48f..7bf5b15 100644
--- a/fpm/src/fpm_compiler.f90
+++ b/fpm/src/fpm_compiler.f90
@@ -228,6 +228,9 @@ character(len=:),allocatable :: module_path_switch
 ! so that values that do not require a space such as -moddir= will work
     model%fortran_compile_flags = fflags//' '//&
     & module_path_switch//join_path(model%output_directory,model%package_name)
+    if (compiler == "nagfor") &
+    & model%fortran_compile_flags = model%fortran_compile_flags//' '//&
+    & "-I "//join_path(model%output_directory,model%package_name)
 
 end subroutine add_compile_flag_defaults
 
```

I know TOML-Fortran and M_CLI2 are compiling with NAG, therefore I gave compiling *fpm* itself a try, of course it does not work, fixing the NAG specific issues is something for another PR however.

More importantly, it might be worth to have all compilers add an include flag for their module output directory as well, in case of `gfortran` it won't do any harm.

Reply:
> Was hoping for a lively discussion on what options to add for some of the compilers I do not have access to, but have not seen much. 

I would not worry about having every compiler setting populated for this PR. This PR sets up the framework; additional compiler support can be added by those with access to such compilers in separate PRs. I'm happy for this PR to go forward as is.

@awvwgk's suggestion for the explicit include flag is a good idea.

Reply:
OK.  So based on that I made a slight change that incorporates that where fflags is the optional flags that a user might override and mandatory is the flags you have to add for ffpm` to work.   It adds some redundancy but solves the problem with nagfor and allows more easily for compilers that do not use -I for locating modules although from the list that seems to be very few anymore.

@awvwgk and @everythingfunctional  do these look like reasonable values for nagfor?  I did some cursory testing with gfortran, ifort, and nvfortran and it does not look like it introduced any problems;  but just guessing at the `fflags` values based on the NAG webpage document.   It would be nice if you can actually try them but as @LKedward said getting optimal values can wait.  I would change them if they actually fail!

Reply:
Works in principle with `nagfor`, I successfully compiled `M_CLI2` with your branch (both debug and release). The main issue with NAG is that a standard violation is always considered a fatal error and is outright rejected by the compiler. So most packages won't compile with `nagfor` anyway because they are not considered standard compliant.

Reply:
@urbanjost , I have not had any luck getting either version of fpm to run on the system I have access to with nag installed. But the flags you've suggested look good to me based on the documentation. As @awvwgk suggested, I think adding the module output directory to the include flags is the simplest way to go.